src/bits/bit_array.rs

```diff
@@ -0,0 +1,181 @@
+use super::{BitArray, BooleanOps};
+use bitvec::prelude::{BitArr, Lsb0};
+use std::{
+    fmt::Debug,
+    ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXor, BitXorAssign, Index, Not},
+};
+
+/// Bit store size of `XorSecretShare` in bytes.
+const BIT_STORE_SIZE_IN_BYTES: usize = 4;
+type U8_4 = BitArr!(for BIT_STORE_SIZE_IN_BYTES * 8, in u8, Lsb0);
+
+/// 32-bit array of bits. Similar to `u32`, but supports boolean algebra, and
+/// provides access to individual bits via index.
+///
+/// Bits are stored in the Little-Endian format. Accessing the first element
+/// like `s[0]` will return the LSB.
+///
+/// ## Example
+/// ```
+/// use raw_ipa::bits::BitArray32;
+///
+/// let s = BitArray32::from(2_u128);
+/// let b0 = s[0];
+/// let b1 = s[1];
+///
+/// assert_eq!(b0, 0);
+/// assert_eq!(b1, 1);
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BitArray32(U8_4);
+
+impl BitArray for BitArray32 {
+    const SIZE_IN_BYTES: usize = BIT_STORE_SIZE_IN_BYTES;
+
+    #[allow(dead_code)]
+    const ZERO: Self = Self(U8_4::ZERO);
+}
+
+impl BooleanOps for BitArray32 {}
+
+impl BitAnd for BitArray32 {
+    type Output = Self;
+    fn bitand(self, rhs: Self) -> Self::Output {
+        Self(self.0 & rhs.0)
+    }
+}
+
+impl BitAndAssign for BitArray32 {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self.0.as_mut_bitslice() &= rhs.0;
+    }
+}
+
+impl BitOr for BitArray32 {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self(self.0 | rhs.0)
+    }
+}
+
+impl BitOrAssign for BitArray32 {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self.0.as_mut_bitslice() |= rhs.0;
+    }
+}
+
+impl BitXor for BitArray32 {
+    type Output = Self;
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        Self(self.0 ^ rhs.0)
+    }
+}
+
+impl BitXorAssign for BitArray32 {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        *self.0.as_mut_bitslice() ^= rhs.0;
+    }
+}
+
+impl Not for BitArray32 {
+    type Output = Self;
+    fn not(self) -> Self::Output {
+        Self(!self.0)
+    }
+}
+
+impl<T: Into<u128>> From<T> for BitArray32 {
+    fn from(v: T) -> Self {
+        Self(U8_4::new(
+            v.into().to_le_bytes()[0..<Self as BitArray>::SIZE_IN_BYTES]
```

Comment:
silently truncating higher-order bits may be surprising for the callers. as I said above, probably better to provide infallible implementation for `u32` and fallible for anything else. It causes clutter at the callsites but it is a good thing in this case imo

Reply:
for fields I believe we don't saturate at Field::PRIME bounds which is what this implementation does. It may also be an option to provide overflowing `From` implementation (similar to `Field`)

Reply:
I think the name `BitArray32` makes it clear that it'll only store the lower 32 bits, similar to `Fp31` or `Fp32BitPrime` accepting `u128`? As in the example above, it helps reducing `try_from().unwrap()` and `&` masks.

With this implementation, `BitArray32::from(1 << 32)` will overflow and yields 0. Correct me if I'm wrong, but I thought saturating will yield the closest valid value, which in this case is 32 1's?

Reply:
iirc, fields have wrapping semantic when converted from the input large than their prime. Apologies for confusion, I meant truncating add here, not saturating. But behavior is still different from what fields do. Besides I think wrapping behavior is pretty much standard for fields, while for bit arrays and arrays in general not so much. 

I worry about truncating semantic because it is lossy and allowing such conversions may lead to subtle issues that will be hard to discover. For protocol code, I think `try_from` is the right choice - if we never convert anything more than 64 bits into bit arrays, then unwrap is (almost) free. but if we have a bug in our code that produces larger than 64 bit items, we will get a panic 

for unit tests I support providing whatever semantic you find convenient when writing them as long as it panics when you do something bad and not visible for production code. `#[cfg(test)]` attribute helps with it.

Reply:
I do understand your concern of not being able to catch subtle issues , but not sure if I would call this conversion lossy. To me, this is masking (= truncating) then converting, which is the same as this code I showed above.

```
    let xor_share = XorReplicated::new(
        u64::try_from(b_bits_left & u128::from(u64::MAX)).unwrap(),
        u64::try_from(b_bits_right & u128::from(u64::MAX)).unwrap(),
    );
```

Since PRSS and fields output u128 values, masking at call sites is what we'll end up with even if we implement `try_from` for bit arrays. We should definitely create a safety net to catch unintended issues, but I don't think fallible conversion here works.

Also, I think wrapping behavior is standard for fields when executing arithmetics, but converting from larger integer types is programmatic, not mathematical. As for bit array conversions, I think masking is pretty standard too. Do you think it'll be better if we create a different function like `mask_from`?

Reply:
Per offline discussion with @akoshelev, I implemented a fallible `TryFrom<u128>` and added `truncate_from`. `truncate_from` should make it more obvious to callers that it will truncate higher-order bits and will lose information.